File: r-mos-backend/app/schemas/report.py
"""
Report相关Pydantic Schema（V2.3完整版）
"""
from pydantic import BaseModel, Field
from typing import Dict, List, Optional, Any
from datetime import datetime


class ScoreBreakdown(BaseModel):
    """评分细分"""
    professionalism: float = Field(..., description="专业性得分（0-25）")
    compliance: float = Field(..., description="规范性得分（0-25）")
    efficiency: float = Field(..., description="效率得分（0-25）")
    safety: float = Field(..., description="安全性得分（0-25）")


class StepScore(BaseModel):
    """步骤得分"""
    step_index: int
    step_title: str
    score: float
    max_score: float
    deductions: List[Dict[str, Any]]
    remarks: Optional[str] = None


class TaskReport(BaseModel):
    """任务报告"""
    task_id: int
    task_title: str
    sop_name: Optional[str] = Field(None, description="SOP名称（可能为NULL）")
    user_id: Optional[int]
    started_at: datetime
    completed_at: datetime
    total_duration_seconds: int
    expected_duration_seconds: Optional[int]
    final_score: float
    pass_score: float
    is_passed: bool
    score_breakdown: ScoreBreakdown
    step_scores: List[StepScore]
    total_steps: int
    completed_steps: int
    skipped_steps: int
    error_count: int
    recommendations: List[str]
    generated_at: datetime


# ============ P2-2: LLM Evaluation Report Schema ============

class LLMEvaluationSection(BaseModel):
    """LLM 生成的评估报告章节"""
    summary: str = Field(..., description="总体评估摘要")
    strengths: List[str] = Field(..., description="学员优势分析")
    improvement_areas: List[str] = Field(..., description="待改进领域")
    root_cause_analysis: str = Field(..., description="问题根本原因分析")
    personalized_suggestions: List[str] = Field(..., description="个性化学习建议")
    next_learning_plan: str = Field(..., description="下一步学习计划")


# ============ P2-3: Peer Comparison Schema ============

class PeerComparisonSection(BaseModel):
    """同伴对比部分"""
    student_level: str = Field(..., description="学员级别")
    group_stats: Dict[str, Any] = Field(..., description="同级别群体统计")
    student_stats: Dict[str, Any] = Field(..., description="学员个人统计")
    comparison: Dict[str, Any] = Field(..., description="对比结果")


class LLMEvaluationReport(BaseModel):
    """LLM 增强的评估报告 - P2-2

    在基础 TaskReport 基础上增加 LLM 生成的叙述性内容：
    - 原因分析
    - 个性化建议
    - 下一步学习计划
    """
    # 基础报告数据
    task_id: int
    task_title: str
    sop_name: Optional[str] = None
    user_id: Optional[int]
    started_at: datetime
    completed_at: datetime
    total_duration_seconds: int
    expected_duration_seconds: Optional[int]
    final_score: float
    pass_score: float
    is_passed: bool
    score_breakdown: ScoreBreakdown
    step_scores: List[StepScore]
    total_steps: int
    completed_steps: int
    skipped_steps: int
    error_count: int
    recommendations: List[str]
    generated_at: datetime

    # LLM 增强内容
    llm_evaluation: Optional[LLMEvaluationSection] = Field(
        default=None,
        description="LLM 生成的评估内容（需要 LLM 服务可用）"
    )
    llm_provider: Optional[str] = Field(
        default=None,
        description="LLM 提供商（如 openai, anthropic）"
    )
    llm_model: Optional[str] = Field(
        default=None,
        description="使用的 LLM 模型"
    )

    # P2-3: 同伴对比
    peer_comparison: Optional[PeerComparisonSection] = Field(
        default=None,
        description="同伴对比（需要足够数据）"
    )

File: r-mos-frontend/src/types/report.ts
/**
 * Report 相关类型定义
 * 与后端 schemas/report.py 完全对齐 (V2.3)
 */

// 评分细分 - 对齐 ScoreBreakdown
export interface ScoreBreakdown {
    professionalism: number             // 专业性得分（0-25）
    compliance: number                  // 规范性得分（0-25）
    efficiency: number                  // 效率得分（0-25）
    safety: number                      // 安全性得分（0-25）
}

// 步骤得分 - 对齐 StepScore
export interface StepScore {
    step_index: number                  // 步骤索引
    step_title: string                  // 步骤标题
    score: number                       // 得分
    max_score: number                   // 满分
    deductions: Array<{                 // 扣分项
        reason: string
        points: number
        [key: string]: any
    }>
    remarks?: string                    // 备注
}

// 任务报告 - 对齐 TaskReport
export interface TaskReport {
    task_id: number                     // 任务ID
    task_title: string                  // 任务标题
    sop_name?: string                   // SOP名称（可能为NULL）
    user_id?: number                    // 用户ID
    started_at: string                  // 开始时间
    completed_at: string                // 完成时间
    total_duration_seconds: number      // 总用时（秒）
    expected_duration_seconds?: number  // 预期用时（秒）
    final_score: number                 // 最终得分
    pass_score: number                  // 及格分数
    is_passed: boolean                  // 是否通过
    score_breakdown: ScoreBreakdown     // 评分细分
    step_scores: StepScore[]            // 各步骤得分
    total_steps: number                 // 总步骤数
    completed_steps: number             // 已完成步骤数
    skipped_steps: number               // 跳过步骤数
    error_count: number                 // 错误次数
    recommendations: string[]           // 改进建议
    generated_at: string                // 报告生成时间
}

// 报告摘要（用于列表展示）
export interface TaskReportSummary {
    task_id: number
    task_title: string
    completed_at: string
    final_score: number
    is_passed: boolean
    total_duration_seconds: number
}
